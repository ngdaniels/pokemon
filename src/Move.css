.move-detail {
    border: 1px solid black;
    border-radius: 3px;
    padding: 1px 4px;
    margin-bottom: 5px;
}

.move-table-container {
    width: 90%;
    margin-left: 1rem;
    overflow: hidden;
    transition: 0.5s;
}

.move-table {
    border-top: 2px solid #705848;
    width: 100%;
    margin: 1rem 0;
}

.table-header-column {
    width: 50%;
    padding: 1rem 0;
}

.move-explanation {
    max-height: 10rem;
    overflow: auto;
    font-size: small;
    line-height: 125%;
}

.arrow {
    float: left;
    transition: 500ms;
}