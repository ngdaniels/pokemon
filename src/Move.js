import React, { Component } from "react";
import './Move.css';
import Type from './Type.js';

class Move extends Component {
    constructor(props) {
        super(props);
        this.state = {
            error: null,
            isLoaded: false,
            move: props.move,
            showDetail: false
        };
    }

    componentDidMount = () => {
        this.mount = true;

        fetch(this.state.move.url)
            .then(res => res.json())
            .then(
                (result) => {
                    if (this.mount) {
                        this.setState({
                            isLoaded: true,
                            moveDetail: result
                        });
                    }
                },
                (error) => {
                    if (this.mount) {
                        this.setState({
                            isLoaded: true,
                            error
                        });
                    }
                }
            )
    }

    componentWillUnmount() {
        this.mount = false;
    }

    convertKebab = (text) => {
        let texts = text.split('-');

        let result = "";

        texts.forEach(text => {
            if (result) {
                result += ' ';
            }
            result += (text.charAt(0).toUpperCase() + text.slice(1));
        });

        return result;
    };

    toggleShowDetail = () => {
        this.setState({
            showDetail: !this.state.showDetail
        });
    };

    render() {
        let { isLoaded, error, moveDetail } = this.state;

        let message;
        if (!isLoaded) {
            message = "Loading Details ...";
        }
        else if (error) {
            message = "Cannot Load Details";
        }

        if (!isLoaded || error) {
            return (
                <div className="move-detail">
                    <div style={{ lineHeight: '3rem' }} onClick={this.toggleShowDetail}>
                        <span className="arrow" style={{ transform: this.state.showDetail ? 'rotate(90deg)' : 'rotate(0deg)' }}>
                            ►
                            </span>
                        <span className="bold">
                            {this.convertKebab(this.state.move.name)}
                        </span>
                    </div>
                    <div className="move-table-container" style={{ maxHeight: this.state.showDetail ? '15rem' : '0' }}>
                        <hr></hr>
                        { message }
                    </div>
                </div>
            )
        } else {
            return (
                <div className="move-detail">
                    <div style={{ lineHeight: '3rem' }} onClick={this.toggleShowDetail}>
                        <span className="arrow" style={{ transform: this.state.showDetail ? 'rotate(90deg)' : 'rotate(0deg)' }}>
                            ►
                        </span>
                        <span className="bold">
                            {this.convertKebab(moveDetail.name)}
                        </span>
                        <span style={{ float: "right" }}>
                            <Type key={moveDetail.name} type={moveDetail.type.name} />
                        </span>
                    </div>
                    <div className="move-table-container" style={{ maxHeight: this.state.showDetail ? '15rem' : '0' }}>
                        <table className="move-table">
                            <thead >
                                <tr>
                                    <td className="table-header-column">
                                        Power: {moveDetail.power ? moveDetail.power : 'N/A'}
                                    </td>
                                    <td className="table-header-column">
                                        PP: {moveDetail.pp}
                                    </td>
                                </tr>
                            </thead>
                            <tbody>
                                <tr>
                                    <td colSpan="2">
                                        <div className="move-explanation">
                                            {moveDetail.effect_entries[0].effect}
                                        </div>
                                    </td>
                                </tr>
                            </tbody>
                        </table>
                    </div>

                </div>
            )
        }
    }
}

export default Move;