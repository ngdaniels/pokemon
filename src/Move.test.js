import { render, screen } from '@testing-library/react';
import Move from './Move';

describe('Move', () => {
	test('Move component before move detail is fetched', () => {
        let { asFragment } = render(<Move move={{ name: "teleport", url: "https://pokeapi.co/api/v2/move/100/" }}/>);
        expect(asFragment()).toMatchSnapshot();
    });
    
    test('Move component after move detail is fetched', async () => {
        let { asFragment } = render(<Move move={{ name: "teleport", url: "https://pokeapi.co/api/v2/move/100/" }}/>);

        let movePower = await screen.findByText("Power", {exact: false});
        expect(movePower).toBeInTheDocument();

        expect(asFragment()).toMatchSnapshot();
	});
});