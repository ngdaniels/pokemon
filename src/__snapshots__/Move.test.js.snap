// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Move Move component after move detail is fetched 1`] = `
<DocumentFragment>
  <div
    class="move-detail"
  >
    <span
      style="float: right;"
    >
      <button
        class="css-xsck4j"
      >
        PSYCHIC
      </button>
    </span>
    <p
      class="bold"
    >
      Teleport
    </p>
    <hr />
    <table
      class="move-table"
    >
      <thead>
        <tr>
          <td
            colspan="2"
          >
            Does nothing.  Wild battles end immediately.
          </td>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td
            style="width: 50%; padding: 1rem 0px;"
          >
            Power: N/A
          </td>
          <td
            style="width: 50%; padding: 1rem 0px;"
          >
            PP: 20
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;

exports[`Move Move component before move detail is fetched 1`] = `
<DocumentFragment>
  <div
    class="move-detail"
  >
    <p
      class="bold"
    >
      Teleport
    </p>
  </div>
</DocumentFragment>
`;
